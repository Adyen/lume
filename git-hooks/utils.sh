#!/bin/sh

RED="\033[1;31m"
GREEN="\033[1;32m"
COLOR_RESET="\033[0m"
