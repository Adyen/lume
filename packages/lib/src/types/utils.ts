export enum Colors {
  Skyblue = 'skyblue',
  Royalblue = 'royalblue',
  Violet = 'violet',
  Darkteal = 'darkteal',
  Gold = 'gold',
}

export enum SequentialColors {
  Skyblue10 = 'skyblue-10',
  Skyblue20 = 'skyblue-20',
  Skyblue30 = 'skyblue-30',
  Skyblue40 = 'skyblue-40',
  Skyblue50 = 'skyblue-50',
  Skyblue60 = 'skyblue-60',
  Skyblue70 = 'skyblue-70',
  Skyblue80 = 'skyblue-80',
  Skyblue90 = 'skyblue-90',
  Skyblue100 = 'skyblue-100',

  Royalblue10 = 'royalblue-10',
  Royalblue20 = 'royalblue-20',
  Royalblue30 = 'royalblue-30',
  Royalblue40 = 'royalblue-40',
  Royalblue50 = 'royalblue-50',
  Royalblue60 = 'royalblue-60',
  Royalblue70 = 'royalblue-70',
  Royalblue80 = 'royalblue-80',
  Royalblue90 = 'royalblue-90',
  Royalblue100 = 'royalblue-100',

  Violet10 = 'violet-10',
  Violet20 = 'violet-20',
  Violet30 = 'violet-30',
  Violet40 = 'violet-40',
  Violet50 = 'violet-50',
  Violet60 = 'violet-60',
  Violet70 = 'violet-70',
  Violet80 = 'violet-80',
  Violet90 = 'violet-90',
  Violet100 = 'violet-100',

  Darkteal10 = 'darkteal-10',
  Darkteal20 = 'darkteal-20',
  Darkteal30 = 'darkteal-30',
  Darkteal40 = 'darkteal-40',
  Darkteal50 = 'darkteal-50',
  Darkteal60 = 'darkteal-60',
  Darkteal70 = 'darkteal-70',
  Darkteal80 = 'darkteal-80',
  Darkteal90 = 'darkteal-90',
  Darkteal100 = 'darkteal-100',

  Gold10 = 'gold-10',
  Gold20 = 'gold-20',
  Gold30 = 'gold-30',
  Gold40 = 'gold-40',
  Gold50 = 'gold-50',
  Gold60 = 'gold-60',
  Gold70 = 'gold-70',
  Gold80 = 'gold-80',
  Gold90 = 'gold-90',
  Gold100 = 'gold-100',
}

export enum DivergentColors {
  SkyTeal = 'skyblue-darkteal',
  VioletRoyal = 'violet-royalblue',
  VioletTeal = 'violet-darkteal',
  GoldTeal = 'gold-darkteal',
  GoldRoyal = 'gold-royalblue',
}

export enum OtherColors {
  White = 'white',
  Black = 'black',
  Green = 'green',
  Orange = 'orange',
  Red = 'red',
  Grey = 'grey',
  Grey10 = 'grey-10',
  Grey20 = 'grey-20',
  Grey30 = 'grey-30',
  Grey40 = 'grey-40',
  Grey50 = 'grey-50',
  Grey60 = 'grey-60',
  Grey70 = 'grey-70',
  Grey80 = 'grey-80',
  Grey90 = 'grey-90',
  Grey100 = 'grey-100',
}

export enum LegacyColors {
  '01' = '01',
  '02' = '02',
  '03' = '03',
  '04' = '04',
  '05' = '05',
  '06' = '06',
  '07' = '07',
  '08' = '08',
  '09' = '09',
}

export type Color = Colors | DivergentColors | LegacyColors | OtherColors;

export type Margins =
  | 'auto'
  | {
      top?: number | 'auto';
      right?: number | 'auto';
      bottom?: number | 'auto';
      left?: number | 'auto';
    };

export type InternalMargins = {
  top: number;
  right: number;
  bottom: number;
  left: number;
};

export type Orientation = 'horizontal' | 'vertical';

export type BarType = 'single' | 'grouped' | 'stacked';

export type BarChartVariant = `${BarType}-bar`;

export type BarLineChartType = BarChartVariant | 'line';

export type AllChartType = 'bar' | BarLineChartType | 'alluvial' | 'sparkline';

export type ChartTypeWithoutVariant = Exclude<AllChartType, BarChartVariant>;
