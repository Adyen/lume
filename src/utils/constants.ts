export enum Colors {
  Skyblue = 'skyblue',
  Royalblue = 'royalblue',
  Violet = 'violet',
  Darkteal = 'darkteal',
  Gold = 'gold',
}

export enum DivergentColors {
  SkyTeal = 'skyblue-darkteal',
  VioletRoyal = 'violet-royalblue',
  VioletTeal = 'violet-darkteal',
  GoldTeal = 'gold-darkteal',
  GoldRoyal = 'gold-royalblue',
}

export enum OtherColors {
  Grey = 'grey',
  Green = 'green',
  Orange = 'orange',
  Red = 'red',
}

export enum LegacyColors {
  '01' = '01',
  '02' = '02',
  '03' = '03',
  '04' = '04',
  '05' = '05',
  '06' = '06',
  '07' = '07',
  '08' = '08',
  '09' = '09',
}

export type Color = Colors | DivergentColors | LegacyColors;

export type Margins = {
  top: number;
  right: number;
  bottom: number;
  left: number;
};

export type Orientation = 'horizontal' | 'vertical';

export const ORIENTATIONS: Record<string, Orientation> = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical',
};

export type BarType = 'single' | 'grouped' | 'stacked';

export const BAR_TYPES: Record<string, BarType> = {
  SINGLE: 'single',
  GROUPED: 'grouped',
  STACKED: 'stacked',
};

export const DEFAULT_MARGINS = {
  VERTICAL: {
    top: 16,
    right: 0,
    bottom: 28,
    left: 24,
  } as Margins,
  HORIZONTAL: {
    top: 16,
    right: 0,
    bottom: 28,
    left: 24,
  } as Margins,
};

export const NO_DATA = 'No data';

export const BAR_HEIGHT = 20;

export const PADDING_VERTICAL = 0.33; // space between bars is 1/2 of a bar's width
export const PADDING_HORIZONTAL = 0.5; // space between bars is a bar's width

// Default radius for the tooltip anchor circle.
// Adding a negligible radius for tooltip anchor, as firefox doesn't respect tooltip element positioning without the circle having actual radius.
export const TOOLTIP_ANCHOR_RADIUS = 1;

export const TOOLTIP_POSITIONS = [
  'auto',
  'auto-start',
  'auto-end',
  'top',
  'top-start',
  'top-end',
  'right',
  'right-start',
  'right-end',
  'bottom',
  'bottom-start',
  'bottom-end',
  'left',
  'left-start',
  'left-end',
] as const; // so that type is not string[]
