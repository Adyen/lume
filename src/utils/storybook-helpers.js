import { boolean, number, select, text } from '@storybook/addon-knobs';

const AXIS_GROUP = 'Axis';
const SIZE_GROUP = 'Size';

const TICK_FORMATS = {
  Null: null,
  Percentage: '~p',
};

/**
 * Provides knob properties to control a chart container size.
 */
export const SizeKnobsMixin = (defaultWidth, defaultHeight) => ({
  width: {
    type: Number,
    default: number('Width', defaultWidth || 540, {}, SIZE_GROUP),
  },
  height: {
    type: Number,
    default: number('Height', defaultHeight || 320, {}, SIZE_GROUP),
  },
});

/**
 * Provides knob properties to control axis options.
 * @param {('x'|'y')} type The axis type.
 */
export const AxisOptionsMixin = (type, isHorizontal = false) => {
  // Dev note: for the current `@storybook/addon-knobs` version (5.0.0) knobs need to have different names
  // even if their groupIds are different. This is was fixed on version [5.2.0](https://github.com/storybookjs/storybook/releases/tag/v5.2.0-alpha.36)
  const groupId = `${type} ` + AXIS_GROUP;
  const defaultGridLines =
    (type === 'y' && !isHorizontal) || (type === 'x' && isHorizontal);
  const prefix = (str) => type + str;

  return {
    [prefix('withLabel')]: {
      type: Boolean,
      default: boolean(type + ' With label', true, groupId),
    },
    [prefix('label')]: {
      type: String,
      default: text(type + ' Lines', '', groupId),
    },
    [prefix('gridLines')]: {
      type: Boolean,
      default: boolean(type + ' Grid lines', defaultGridLines, groupId),
    },
    [prefix('showTicks')]: {
      type: Boolean,
      default: boolean(type + ' Show ticks', true, groupId),
    },
    [prefix('tickCount')]: {
      type: Number,
      default: number(type + ' Tick count', undefined, {}, groupId),
    },
    [prefix('tickFormat')]: {
      type: String,
      default: select(
        type + ' Tick format',
        TICK_FORMATS,
        TICK_FORMATS.Null,
        groupId
      ),
    },
    [prefix('tickPadding')]: {
      type: Number,
      default: number(type + ' Tick padding', 8, {}, groupId),
    },
  };
};

/**
 * Converts prefixed knob properties into regular axis options.
 * This is required since `@storybook/addon-knobs` can't handle same name properties.
 * @param {Object} props Story component properties.
 * @param {('x'|'y')} type The axis type.
 * @returns {Object} An axis options object.
 */
export const convertAxisPropsIntoOptions = (props, type) => {
  const keys = Object.keys(props).filter((key) => key.startsWith(type));
  return keys.reduce((options, key) => {
    options[key.substring(1)] = props[key];
    return options;
  }, {});
};
