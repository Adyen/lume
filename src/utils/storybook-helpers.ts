const AXIS_GROUP = 'Axis';
const SIZE_GROUP = 'Size';

const TICK_FORMATS = {
  Null: null,
  Percentage: '~p',
};

/**
 * Provides knob properties to control a chart container size.
 */

export const withSizeArgTypes = () => ({
  width: {
    control: { type: 'number', step: 10 },
    description: 'Chart width value (px).',
  },
  height: {
    control: { type: 'number', step: 10 },
    description: 'Chart height value (px).',
  },
});

export const withSizeArgs = (
  defaultWidth?: number,
  defaultHeight?: number
) => ({
  width: defaultWidth || 540,
  height: defaultHeight || 360,
});
