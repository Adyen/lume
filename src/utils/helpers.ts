import { isRef, Ref } from 'vue';
import { ScaleBand } from 'd3-scale';
import { SankeyNode } from 'd3-sankey';

import {
  SankeyNodeAdditionalProperties,
  SankeyLinkAdditionalProperties,
} from '@/types/alluvial';
import { Data, DatasetValueObject } from '@/types/dataset';
import { Scale } from '@/composables/scales';

/**
 * Returns an array with all numeric values present in a chart's `data` (array of datasets).
 * @param {Data<DatasetValueObject>} data A computed array of datasets.
 * @returns {Array<number>} An array of all numeric values.
 */
export function flatValues(data: Data<DatasetValueObject>): Array<number> {
  return (
    data
      ?.map((dataset) =>
        dataset.values.map((datasetValue) => datasetValue?.value).flat()
      )
      .flat() || []
  );
}

/**
 * Checks if provided data has only 1 dataset.
 * @param {Data} data The data prop
 * @returns {boolean} True if valid single dataset data
 */
export function singleDatasetValidator(data: Data): boolean {
  return data.length === 1;
}

/**
 * Checks if provided item is an object.
 * @param {unknown} item
 * @returns {boolean} True if item is an object.
 */
export function isObject(item: unknown): item is Record<string, unknown> {
  return !!item && typeof item === 'object' && !Array.isArray(item);
}

/**
 * Deep merge n objects.
 * Clones both target and sources so that they don't get mutated.
 *
 * @param {Record<string, unknown>} target Object to merge to.
 * @param {Array<Record<string, unknown>>} source Object(s) to merge into target.
 */
export function mergeDeep(
  target: Record<string, unknown>,
  source: Record<string, unknown>
) {
  const object = structuredClone(target);

  Object.entries(source).forEach(([key, value]) => {
    object[key] =
      isObject(target[key]) && isObject(value)
        ? mergeDeep(target[key] as Record<string, unknown>, value)
        : value;
  });

  return object;
}

/**
 * Checks if provided item is a DatasetValueObject.
 * @param {unknown} object
 * @returns {boolean} True if item is a DatasetValueObject.
 */
export function isDatasetValueObject(
  object: unknown
): object is DatasetValueObject {
  return (
    object != null &&
    isObject(object) &&
    'value' in (object as unknown as DatasetValueObject)
  );
}

export function isBandScale(scale: Scale): scale is ScaleBand<string | number> {
  return (scale as ScaleBand<string | number>)?.bandwidth !== undefined;
}

export function getScaleStep(scale: Scale) {
  if (isBandScale(scale)) return scale.bandwidth();
  return Math.max(...scale.range()) / Math.max(...scale.domain());
}

const validateGetHighestValueArguments = (
  data: Data<DatasetValueObject>,
  index: number
) => {
  if (!data) {
    throw new Error('Data is not a valid array');
  }

  if (data.length === 0) {
    throw new Error('Cannot get highest value from empty array');
  }

  data.forEach(({ values }) => {
    if (values.length - 1 < index) {
      throw new Error('Index exceeds length of at least one of the datasets');
    }
  });
};

/**
 * Returns the highest value of all provided datasets in a given index.
 *
 * @param {Data} data The data to use.
 * @param {number} index THe index to check for.
 * @returns {number} The highest value.
 */
export function getHighestValue(
  data: Data<DatasetValueObject>,
  index: number
): number {
  validateGetHighestValueArguments(data, index);

  return (
    data.reduce((max, point) =>
      (max.values[index]?.value ?? 0) > (point.values[index]?.value ?? 0)
        ? max
        : point
    ).values[index]?.value || 0
  );
}

/**
 * Returns an empty array with length equal to that of the dataset.
 *
 * @param data A data group.
 * @returns An empty array.
 */
export function getEmptyArrayFromData(data: Data | Ref<Data>) {
  const dataArray = isRef(data) ? data.value : data;
  if (!dataArray) {
    throw new Error('No empty array can be created from specified data value');
  }

  // Use max length in case datasets have different lengths
  const maxLength = Math.max(
    ...dataArray.map((dataset) => dataset.values.length),
    0
  );

  return Array(...Array(maxLength));
}

/**
 * Returns the length of a scale domain.
 *
 * @param scale The scale to get the length from.
 * @returns The domain length.
 */
export function getDomainLength(scale: Scale): number {
  return isBandScale(scale)
    ? scale.domain().length
    : Math.max(...scale.domain());
}

/**
 * Generates an interpolator from `start` to `end`, this is equivalent to d3-interpolate interpolateRound function.
 * @param start {number}
 * @param end {number}
 */
export function interpolateRound(start: number, end: number) {
  return (t: number) => Math.round(start * (1 - t) + end * t);
}

/**
 * Gets the ID of an alluvial node.
 * @param node An alluvial node.
 * @returns The node ID.
 */
export function getAlluvialNodeId(
  node:
    | string
    | number
    | SankeyNode<SankeyNodeAdditionalProperties, SankeyLinkAdditionalProperties>
): number | string {
  if (typeof node === 'string' || typeof node === 'number') return node;
  return node.id ?? node.label;
}

/**
 * Checks if a given node or link is being hovered.
 * @param id The node/link ID.
 * @param highlightedIds An array of IDs currently hovered.
 * @returns True if node is being hovered.
 */
export function isNodeOrLinkFaded(id: string, highlightedIds: Array<string>) {
  return highlightedIds.length && highlightedIds.indexOf(id) === -1;
}
