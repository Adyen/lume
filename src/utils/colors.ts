import { ColorPalette } from '@/types/dataset';
import { Color, Colors, DivergentColors, LegacyColors } from '@/constants';
import { shiftItems } from './helpers';

const DEFAULT_PALETTE = ColorPalette.Categorical;
const DEFAULT_COLOR = Colors.Skyblue;
const DEFAULT_DIVERGENT_COLOR = DivergentColors.SkyTeal;

function getDefaultColorByPalette(palette: ColorPalette) {
  if (palette === ColorPalette.Divergent) return DEFAULT_DIVERGENT_COLOR;
  return DEFAULT_COLOR;
}

function isLegacy(color: Color) {
  return Object.values(LegacyColors).includes(color as LegacyColors);
}

function validateColor(color: Color | null, palette: ColorPalette) {
  if (palette === ColorPalette.Divergent) {
    return Object.values(DivergentColors).includes(color as DivergentColors);
  }
  return Object.values({ ...Colors, ...LegacyColors }).includes(
    color as Colors | LegacyColors
  );
}

function getCategoricalColor(color: Colors, index: number) {
  // When computing the categorical color, we need to move items in the colors enum to
  //   be able to correctly map them to the datasets based on the provided index.
  // For instance, if the user says the first color should be `royalblue`, then we
  //   move items in the enum so that index 0 is now `royalblue` (and `skyblue` is last).
  const colorIndex = Object.values(Colors).indexOf(color);
  return shiftItems(Object.values(Colors), colorIndex)[index] as Colors;
}

function getSequentialColor(color: Colors, index: number) {
  const value = 50 + index * 10;
  return `${color}-${value}`;
}

function getDivergentColor(color: DivergentColors, index: number) {
  const value = (index + 1) * 10;
  return `${color}-${value}`;
}

function getColorByPalette(color: Color, palette: ColorPalette, index: number) {
  switch (palette) {
    default:
    case ColorPalette.Categorical:
      return isLegacy(color)
        ? color
        : getCategoricalColor(color as Colors, index);
    case ColorPalette.Sequential:
      return getSequentialColor(color as Colors, index);
    case ColorPalette.Divergent:
      return getDivergentColor(color as DivergentColors, index);
  }
}

export function computeColor(
  datasetColor: Color,
  color: Color,
  palette: ColorPalette = DEFAULT_PALETTE,
  index: number
) {
  const isDatasetColorValid = validateColor(datasetColor, palette);
  if (isDatasetColorValid) {
    return getColorByPalette(datasetColor, palette, index);
  }

  const isColorValid = validateColor(color, palette);
  if (isColorValid) return getColorByPalette(color, palette, index);

  return getColorByPalette(getDefaultColorByPalette(palette), palette, index);
}

export function getDataTypeClass(type: ColorPalette = DEFAULT_PALETTE) {
  return `lume-data-type--${type}`;
}
