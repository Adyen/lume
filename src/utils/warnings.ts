import { isProd } from './helpers';

export enum Warnings {
  DatasetLength = 'Our guidelines highly recommend not having more than 5 datasets in one chart.\nLearn more: https://lume.design/guidelines/datasets',
  ColorLoop = 'Colors will loop around after the 5th dataset. You can prevent this by setting specific colors for your datasets.',
}

const PREFIX = '[lume] ';

export function warn(warning: Warnings, devOnly = true) {
  if (devOnly && isProd()) return;
  console.warn(PREFIX + 'Warning: ' + warning);
}
