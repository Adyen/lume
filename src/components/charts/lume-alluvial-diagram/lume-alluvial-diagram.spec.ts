import { mount } from '@vue/test-utils';

import LumeAlluvialDiagram from './lume-alluvial-diagram.vue';

import { BaseTestSuite } from '@test/unit/reusable.test';
import { generateData } from '@test/unit/utils/data-generator';

import DATASETS from '@/docs/storybook-data/alluvial-data';

const AlluvialDiagramTestSuiteFactory = (data: any = DATASETS.Basic.data) =>
  new BaseTestSuite(LumeAlluvialDiagram, { data });
const [baseData] = generateData();

describe('lume-alluvial-diagram.vue', () => {
  let originalError = null;
  const stageWithData = (values) => {
    const data = [
      {
        ...baseData,
        values,
      },
    ];

    const testSuite = AlluvialDiagramTestSuiteFactory(data);

    const wrapper = testSuite.run().wrapper;
    const el = wrapper.find('[data-j-alluvial-diagram]');
    expect(el.exists()).toBeTruthy();
  };

  beforeAll(() => {
    // Suppress error output to the console
    originalError = console.error;
    console.error = vi.fn();
  });

  afterAll(() => {
    // Restore default error output setting
    console.error = originalError;
  });

  test('mounts component and sets prop values', async () => {
    const testSuite = AlluvialDiagramTestSuiteFactory();

    const wrapper = testSuite.run().wrapper;
    const el = wrapper.find('[data-j-alluvial-diagram]');
    expect(el.exists()).toBeTruthy();
    expect(el.find('[data-j-alluvial-group]').exists()).toBeTruthy();
    expect(
      el.find('[data-j-alluvial-group__ghost-path]').exists()
    ).toBeTruthy();
    expect(el.find('[data-j-alluvial-group__path]').exists()).toBeTruthy();
    expect(el.find('[data-j-alluvial-group__node-block]').exists()).toBeFalsy();
  });

  test('mounts component and sets prop values using random data', async () => {
    const data = generateData();
    const testSuite = AlluvialDiagramTestSuiteFactory(data);

    const wrapper = testSuite.run().wrapper;
    const el = wrapper.find('[data-j-alluvial-diagram]');
    expect(el.exists()).toBeTruthy();
    expect(el.find('[data-j-alluvial-group]').exists()).toBeTruthy();
    expect(
      el.find('[data-j-alluvial-group__ghost-path]').exists()
    ).toBeTruthy();
    expect(el.find('[data-j-alluvial-group__path]').exists()).toBeTruthy();
    expect(el.find('[data-j-alluvial-group__node-block]').exists()).toBeFalsy();
  });

  test('should throw error in case of a dataset with circular links', async () => {
    const values = [
      {
        label: 'A',
        color: '01',
        value: 'A',
        targets: [{ node: 'B', value: 50 }],
      },
      {
        label: 'B',
        color: '02',
        value: 'B',
        targets: [{ node: 'A', value: 50 }],
      },
    ];

    const data = [
      {
        ...baseData,
        values,
      },
    ];

    const t = () =>
      mount(LumeAlluvialDiagram, {
        propsData: {
          data,
        },
      });

    expect(t).toThrow('circular link');
  });

  test('should be able to handle an empty dataset', async () => {
    const values = [];

    const data = [
      {
        ...baseData,
        values,
      },
    ];

    const t = () =>
      mount(LumeAlluvialDiagram, {
        propsData: {
          data,
        },
      });

    expect(t).toThrow('Invalid array length');
  });

  test('should be able to handle an dataset with a single node', async () => {
    const values = [{ label: 'A', color: '01', value: 'A', targets: [] }];
    stageWithData(values);
  });

  test('should be able to handle an dataset with only nodes but no edges', async () => {
    const values = [
      { label: 'A', color: '01', value: 'A', targets: [] },
      { label: 'B', color: '02', value: 'B', targets: [] },
    ];
    stageWithData(values);
  });

  test('should be able to handle an dataset with links to non-existing nodes', async () => {
    const values = [
      { label: 'A', color: '01', value: 'A', targets: [] },
      {
        label: 'B',
        color: '02',
        value: 'B',
        targets: [{ node: 'C', value: 50 }],
      },
    ];
    const data = [
      {
        ...baseData,
        values,
      },
    ];

    const t = () =>
      mount(LumeAlluvialDiagram, {
        propsData: {
          data,
        },
      });

    expect(t).toThrow('missing: C');
  });

  // Test interaction with elements
});
