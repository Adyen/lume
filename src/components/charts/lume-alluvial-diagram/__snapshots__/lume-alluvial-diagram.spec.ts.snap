// Vitest Snapshot v1

exports[`lume-alluvial-diagram.vue > mounts component and sets prop values 1`] = `
<div
  class="lume-chart-container lume-chart-container--minimum"
  data-j-alluvial-diagram=""
  data-j-chart-container=""
  data-j-lume-chart=""
  data-v-8124b354=""
  data-v-e7119ab8=""
>
  <div
    class="lume-chart__header"
    data-v-8124b354=""
    data-v-e7119ab8=""
  >
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <div
        class="lume-chart__controls"
        data-v-8124b354=""
        data-v-e7119ab8=""
      />
    </div>
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <!---->
    </div>
  </div>
  <svg
    class="lume-chart-container__svg"
    data-j-chart-container__root=""
    data-v-8124b354=""
    height="0"
  >
    <g
      class="lume-chart-container__group"
      data-j-chart-container__group=""
      data-v-8124b354=""
      transform="translate(0, 0)"
    >
      <!---->
      <!---->
      <g
        class="lume-alluvial-group"
        containerSize="[object Object]"
        data-j-alluvial-group=""
        data-v-8124b354=""
        data-v-ce3d709d=""
        hoveredIndex="-1"
        transition="true"
      >
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__ghost-path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
      </g>
      <!---->
      <!---->
    </g>
  </svg>
  <!---->
  <!---->
</div>
`;

exports[`lume-alluvial-diagram.vue > mounts component and sets prop values using random data 1`] = `
<div
  class="lume-chart-container lume-chart-container--minimum"
  data-j-alluvial-diagram=""
  data-j-chart-container=""
  data-j-lume-chart=""
  data-v-8124b354=""
  data-v-e7119ab8=""
>
  <div
    class="lume-chart__header"
    data-v-8124b354=""
    data-v-e7119ab8=""
  >
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <div
        class="lume-chart__controls"
        data-v-8124b354=""
        data-v-e7119ab8=""
      />
    </div>
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <!---->
    </div>
  </div>
  <svg
    class="lume-chart-container__svg"
    data-j-chart-container__root=""
    data-v-8124b354=""
    height="0"
  >
    <g
      class="lume-chart-container__group"
      data-j-chart-container__group=""
      data-v-8124b354=""
      transform="translate(0, 0)"
    >
      <!---->
      <!---->
      <g
        class="lume-alluvial-group"
        containerSize="[object Object]"
        data-j-alluvial-group=""
        data-v-8124b354=""
        data-v-ce3d709d=""
        hoveredIndex="-1"
        transition="true"
      >
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__ghost-path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
      </g>
      <!---->
      <!---->
    </g>
  </svg>
  <!---->
  <!---->
</div>
`;

exports[`lume-alluvial-diagram.vue > should be able to handle an dataset with a single node 1`] = `
<div
  class="lume-chart-container lume-chart-container--minimum"
  data-j-alluvial-diagram=""
  data-j-chart-container=""
  data-j-lume-chart=""
  data-v-8124b354=""
  data-v-e7119ab8=""
>
  <div
    class="lume-chart__header"
    data-v-8124b354=""
    data-v-e7119ab8=""
  >
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <div
        class="lume-chart__controls"
        data-v-8124b354=""
        data-v-e7119ab8=""
      />
    </div>
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <!---->
    </div>
  </div>
  <svg
    class="lume-chart-container__svg"
    data-j-chart-container__root=""
    data-v-8124b354=""
    height="0"
  >
    <g
      class="lume-chart-container__group"
      data-j-chart-container__group=""
      data-v-8124b354=""
      transform="translate(0, 0)"
    >
      <!---->
      <!---->
      <g
        class="lume-alluvial-group"
        containerSize="[object Object]"
        data-j-alluvial-group=""
        data-v-8124b354=""
        data-v-ce3d709d=""
        hoveredIndex="-1"
        transition="true"
      >
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__ghost-path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
      </g>
      <!---->
      <!---->
    </g>
  </svg>
  <!---->
  <!---->
</div>
`;

exports[`lume-alluvial-diagram.vue > should be able to handle an dataset with only nodes but no edges 1`] = `
<div
  class="lume-chart-container lume-chart-container--minimum"
  data-j-alluvial-diagram=""
  data-j-chart-container=""
  data-j-lume-chart=""
  data-v-8124b354=""
  data-v-e7119ab8=""
>
  <div
    class="lume-chart__header"
    data-v-8124b354=""
    data-v-e7119ab8=""
  >
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <div
        class="lume-chart__controls"
        data-v-8124b354=""
        data-v-e7119ab8=""
      />
    </div>
    <div
      class="lume-chart__header-section"
      data-v-8124b354=""
      data-v-e7119ab8=""
    >
      <!---->
      <!---->
    </div>
  </div>
  <svg
    class="lume-chart-container__svg"
    data-j-chart-container__root=""
    data-v-8124b354=""
    height="0"
  >
    <g
      class="lume-chart-container__group"
      data-j-chart-container__group=""
      data-v-8124b354=""
      transform="translate(0, 0)"
    >
      <!---->
      <!---->
      <g
        class="lume-alluvial-group"
        containerSize="[object Object]"
        data-j-alluvial-group=""
        data-v-8124b354=""
        data-v-ce3d709d=""
        hoveredIndex="-1"
        transition="true"
      >
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__ghost-path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
        <g
          class="lume-alluvial-path-group"
          data-j-alluvial-group__path=""
          data-v-ce3d709d=""
          data-v-e924c8b9=""
        />
      </g>
      <!---->
      <!---->
    </g>
  </svg>
  <!---->
  <!---->
</div>
`;
